### PR TITLE
Added YUM implementation of OS package build tool

### PR DESCRIPTION
**Pull request by fheinecke**
This PR has several inter-dependent changes:
* Renamed the "build-apt-repos" to "build-os-package-repos"
* Rewrote the tool to use subcommands (i.e. from `go run . <args>` to `go run . <apt/yum> <args>`)
* Implemented YUM repo building per [rfd/0058-package-distribution.md](https://github.com/gravitational/teleport/blob/master/rfd/0058-package-distribution.md)
* Added os-specific "*.repo" files for use with `dnf config-manager` and `yum-config-manager`
* Added redirects on APT and YUM buckets from `index.html` to https://goteleport.com/docs/installation/
* Refactored dronegen to more easily support adding OS package repos to the promotion pipeline
* Added YUM repo building to dronegen
* Added dronegen parallelism support
* Added dronegen resource limit support

Before this PR is merged I need to run migrations for old Teleport versions to populate the prod bucket with old release artifacts. Additionally much of it's functionality is dependent on https://github.com/gravitational/cloud-terraform/pull/701 being merged first.

This PR is ready for code review but not yet ready for merge.

### CODE REVIEW

File: build.assets/tooling/cmd/build-os-package-repos/gpg.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+	err = g.ensureSecretKeyExists()
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to ensure a secret key exists")
+	}
+
+	return g, nil
+}
+
+// The first time GPG is run for a user with any "meaningful" arguments it will
```
Comment (tcsc) [inline build.assets/tooling/cmd/build-os-package-repos/gpg.go:46]: Do you need to use the system PGP? If you're just signing things you might be able to get away with `github.com/ProtonMail/go-crypto/openpgp`. See `teleport-plugins/tooling`

Comment (fheinecke) [reply build.assets/tooling/cmd/build-os-package-repos/gpg.go:46]: Did a bunch of reading/testing and neither `golang.org/x/crypto/openpgp` nor `github.com/ProtonMail/go-crypto/openpgp` support keybox (.kbx) files, which the signing keypair use. There are a number of workarounds here, but the only one that doesn't have fairly large downsides is using the system's GPG binary. Several other projects (sops, aptly) use this approach as well.

**Comment by r0mant** [comment]
@capnspacehook @timothyb89 Could you please take a look at this PR?

File: dronegen/apt.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+func migrateAptPipeline(triggerBranch string, migrationVersions []string) pipeline {
+	return getAptPipelineBuilder().buildMigrateOsPackagePipeline(triggerBranch, migrationVersions)
+}
+
+func getAptPipelineBuilder() *OsPackageToolPipelineBuilder {
+	optpb := NewOsPackageToolPipelineBuilder(
+		"drone-s3-aptrepo-pvc",
+		"golang:1.18.1-bullseye",
```
Comment (capnspacehook) [inline dronegen/apt.go:31]: Any reason Go 18.1 is used for apt and 18.3 is used for rpm?

File: build.assets/tooling/cmd/build-os-package-repos/command_executor.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	cmd := exec.Command(command, args...)
+	logrus.Debugf("Running command \"%s '%s'\"", command, strings.Join(args, "' '"))
+	output, err := cmd.CombinedOutput()
+
+	if output != nil {
+		logrus.Debugf("Command output: %s", string(output))
+	}
+
+	if err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
```
Comment (capnspacehook) [inline build.assets/tooling/cmd/build-os-package-repos/command_executor.go:40]: Nit: use `errors.As` here?

**Review by capnspacehook — APPROVED** [review]

File: dronegen/os_repos.go [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+// Used for one-off migrations of older versions.
+// Use cases include:
+//  * We want to support another OS while providing backwards compatibility
+//  * We want to support another OS version while providing backwards compatibility
+//  * A customer wants to be able to install an older version via APT/YUM even if we
+//      no longer support it
+//  * RPM migrations after new YUM pipeline is done
+func artifactMigrationPipeline() []pipeline {
+	migrationVersions := []string{
+		// These versions were migrated as a part of the new `promoteAptPipeline`
```
Comment (timothyb89) [inline dronegen/os_repos.go:41]: This seems a bit unclear to me given all the entries are commented out. Is it a "add the version temporarily to the list then comment it out once migrated" situation? Is there a procedure we could link to if it comes up in the future? (or does it matter?)

File: dronegen/yum.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+func migrateYumPipeline(triggerBranch string, migrationVersions []string) pipeline {
+	return getYumPipelineBuilder().buildMigrateOsPackagePipeline(triggerBranch, migrationVersions)
+}
+
+func getYumPipelineBuilder() *OsPackageToolPipelineBuilder {
+	optpb := NewOsPackageToolPipelineBuilder(
+		"drone-s3-yumrepo-pvc",
+		"centos:7", // This can probably be bumped but newer versions don't appear to have createrepo for some reason
```
Comment (timothyb89) [inline dronegen/yum.go:34]: That's a bit scary. It looks like CentOS 8 is already EoL and they took the repo offline :/

If I understand things correctly, I don't think this pipeline actually builds any artifacts? If so, maybe we should consider `rockylinux:9` or something? Most distros still package `createrepo_c` / `createrepo-c`, so we could probably use Ubuntu if we wanted.

File: build.assets/tooling/cmd/build-os-package-repos/runners.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+				"zesty",    // 17.04 (EOL)
+				"artful",   // 17.10 (EOL)
+				"bionic",   // 18.04 LTS
+				"cosmic",   // 18.10 (EOL)
+				"disco",    // 19.04 (EOL)
+				"eoan",     // 19.10 (EOL)
+				"focal",    // 20.04 LTS
+				"groovy",   // 20.10 (EOL)
+				"hirsuite", // 21.04 (EOL)
+				"impish",   // 21.10 (EOL on 7/14/22)
```
Comment (timothyb89) [inline build.assets/tooling/cmd/build-os-package-repos/runners.go:63]: nit: looks like this is now EoL

**Review by timothyb89 — APPROVED** [review]
Looks good to me, though with the caveat that I haven't really tested anything. Just a couple small questions!

File: dronegen/os_repos.go (hunk 4 above)
Comment (fheinecke) [reply dronegen/os_repos.go:41]: That's exactly it. I can write up some docs somewhere... but outside of adding a new OS package repo type (like YUM or APK) it's pretty unlikely that we'll need to use this unless a customer complains about an old version being missing. It's really nice to have (greatly simplifies migrations) but will probably be very infrequently used after this PR is completed.

File: dronegen/yum.go (hunk 5 above)
Comment (fheinecke) [reply dronegen/yum.go:34]: Nope, the pipeline does not actually build any artifacts - only YUM repos. I'll bump this to `ubuntu:22.04` which provides `createrepo_c` via the `createrepo-c` package

File: build.assets/tooling/cmd/build-os-package-repos/command_executor.go (hunk 3 above)
Comment (fheinecke) [reply build.assets/tooling/cmd/build-os-package-repos/command_executor.go:40]: Good catch - updated

File: dronegen/apt.go (hunk 2 above)
Comment (fheinecke) [reply dronegen/apt.go:31]: When I wrote the APT portion only 18.1 was available. Updated both to 18.4

File: build.assets/tooling/cmd/build-os-package-repos/runners.go (hunk 6 above)
Comment (fheinecke) [reply build.assets/tooling/cmd/build-os-package-repos/runners.go:63]: Thanks - updated

**Review by r0mant — APPROVED** [review]

**Pull request merged**